modules/javafx.controls/src/main/java/javafx/scene/control/ControlUtils.java

```diff
@@ -166,19 +167,26 @@ public static <S> void updateSelectedIndices(MultipleSelectionModelBase<S> sm, b
             // ensuring that the selectedIndices bitset is correctly updated.
 
             sm.startAtomic();
-            final List<Integer> removed = c.getRemoved().stream()
+
+            final List<Integer> removed = new ArrayList<>(c.getRemovedSize());
+            c.getRemoved().stream()
                     .mapToInt(TablePositionBase::getRow)
                     .distinct()
                     .filter(removeRowFilter)
-                    .boxed()
-                    .peek(sm.selectedIndices::clear)
-                    .collect(Collectors.toList());
+                    .forEach(row -> {
```

Comment:
In the java.util.stream package [docs](https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/util/stream/package-summary.html#SideEffects) it is mentioned that `forEach()` method operates only via side-effects. So do you think we should avoid using `forEach()` here and iterate the generated list separately to clear selected index?

Reply:
I'd say .forEach() is used correctly here, according to docs, it guarantees execution of the side-effects (add to removed list & clear index), just not in any particular order. This way we avoid multiple iteration.

Reply:
Another idea is to use `toList()`, which is a very efficient operation and then iterate over it.

Reply:
> In the java.util.stream package [docs](https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/util/stream/package-summary.html#SideEffects) it is mentioned that `forEach()` method operates only via side-effects. So do you think we should avoid using `forEach()` here and iterate the generated list separately to clear selected index?

`forEach` is used correctly here. From the docs:
> With the exception of terminal operations [forEach](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/stream/Stream.html#forEach(java.util.function.Consumer)) and [forEachOrdered](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/stream/Stream.html#forEachOrdered(java.util.function.Consumer)), side-effects of behavioral parameters may not always be executed when the stream implementation can optimize away the execution of behavioral parameters without affecting the result of the computation.

> Another idea is to use `toList()`, which is a very efficient operation and then iterate over it.

That's still 2 iterations. If the code is not performance-critical it doesn't matter.

Reply:
> That's still 2 iterations.

Yes, but one advantage here:
We currently do `final List<Integer> removed = new ArrayList<>(c.getRemovedSize());`, 
where we allocate a list with a size, that is probably too big since we filter the removed items.
So with `toList`, we at least get back a list with the correct size.
But true, that we technically iterate twice then. It probably does not matter too much.

Reply:
> `forEach` is used correctly here. From the docs:
> 
> > With the exception of terminal operations [forEach](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/stream/Stream.html#forEach(java.util.function.Consumer)) and [forEachOrdered](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/stream/Stream.html#forEachOrdered(java.util.function.Consumer)), side-effects of behavioral parameters may not always be executed when the stream implementation can optimize away the execution of behavioral parameters without affecting the result of the computation.
> 

Thanks for pointing this out.